src/service/xhr-impl.js

```diff
@@ -125,43 +122,43 @@ export class Xhr {
    * @private
    */
   fetch_(input, init) {
+    const ampdocSingle = this.ampdocService.getAmpDoc();
     if (!getMode().test &&
-        this.ampdocSingle_ &&
+        ampdocSingle &&
         Math.random() < 0.01 &&
-        parseUrl(input).origin != this.win.location.origin &&
-        !Services.viewerForDoc(this.ampdocSingle_).hasBeenVisible()) {
-      dev().error('XHR', 'attempted to fetch %s before viewer was visible',
+        parseUrl(input).origin != this.win.location.origin) {
+      dev().error('XHR', 'attempted to fetch URL from different origin',
           input);
     }
-
     dev().assert(typeof input == 'string', 'Only URL supported: %s', input);
     // In particular, Firefox does not tolerate `null` values for
     // `credentials`.
     const creds = init.credentials;
     dev().assert(
         creds === undefined || creds == 'include' || creds == 'omit',
         'Only credentials=include|omit support: %s', creds);
-
-    return this.maybeIntercept_(input, init).then(interceptorResponse => {
-      if (interceptorResponse) {
-        return interceptorResponse;
-      }
-
-      // After this point, both the native `fetch` and the `fetch` polyfill will
-      // expect a native `FormData` object in the `body` property, so the native
-      // `FormData` object needs to be unwrapped.
-      if (isFormDataWrapper(init.body)) {
-        init.body = init.body.getFormData();
-      }
-      // Fallback to xhr polyfill since `fetch` api does not support
-      // responseType = 'document'. We do this so we don't have to do any
-      // parsing and document construction on the UI thread which would be
-      // expensive.
-      if (init.responseType == 'document') {
-        return fetchPolyfill(input, init);
-      }
-      return (this.win.fetch || fetchPolyfill).apply(null, arguments);
-    });
+    return Services.viewerForDoc(ampdocSingle).whenFirstVisible()
+        .then(() => {
+          return this.maybeIntercept_(input, init, ampdocSingle);
```

Comment:
@zhangsu: What are the privacy guarantees of the Gmail interceptor? Does it need us to wait for the doc to be shown?

(Basically, should this be the first call, or should `whenFirstVisible` be first)

Reply:
Thanks for checking with us. Waiting for the email to be shown before firing XHRs sounds good to me since we also want the same level of privacy.

Just curious: what's an example today where XHRs can be fired before the viewer is visible (e.g. in prerender mode)? `amp-list` shouldn't do that because `prerenderAllowed` is false, right?

Reply:
👍, that makes it safer on our side as well.

> what's an example today where XHRs can be fired before the viewer is visible (e.g. in prerender mode)?

Generally, these are bugs. There are exceptions for connecting to the CDN itself (or whatever origin is hosting the page), since they've already received the document request.

> amp-list shouldn't do that because prerenderAllowed is false, right?

Correct. The `prerenderAllowed` is one way to catch them, but we're increasingly getting extensions that need to make requests before layout (ads 🙄, and one other). These make requests in `#buildCallback` or other earlier lifecycle callbacks, and they sometimes forget to wait for visible.